[Bootcamp][pytorch]Add HTTPReaderIterDataPipe and ToBytesIterDataPipe

jg2912 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59816 [Bootcamp][pytorch]Add HTTPReaderIterDataPipe and ToBytesIterDataPipe**

Summary: Add three new DataPipes, 
readlinesfromfile for getting file names to yield text lines
httpreader for getting web file urls to yield streams 
tobytes for getting streams to yield bytes.

Test Plan: new test case TestIterableDataPipeHttp in test/test_datapipes.py. The test initiates a local http server for serving test files. Local tests are ok. 
1. Create 16M files and load from the local server. Each of size 10 Bytes. Purpose is to see if any unreleased OS handles from data pipes. Notes, it takes about 5 hours to run at this configuration.
2. Create a 64GB file and load from the local server. Read is chunked at 8MB. Purpose is to see if it could load a large web file using the data pipes. 
In case of users inadvertently running into a long test case, configurations of both are put to smaller numbers.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29051186](https://our.internmc.facebook.com/intern/diff/D29051186)

Commit:
[Bootcamp][pytorch]Add WebIterDataPipe and ToBytesIterDataPipe to the datapipes.

Summary: Add two new DataPipes, one for getting web file urls to yield streams and one for getting streams to yield bytes.

Test Plan:Add test_web_iterable_datapipe in test/test_datapipes.py. The test initiates a local http server for serving test files. test is ok.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "[Bootcamp][pytorch]Add WebIterDataPipe and ToBytesIterDataPipe to the datapipes."


Summary: Add two new DataPipes, one for getting web file urls to yield streams and one for getting streams to yield bytes.

Test Plan:Add test_web_iterable_datapipe in test/test_datapipes.py. The test initiates a local http server for serving test files. test is ok.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "[Bootcamp][pytorch]Add WebIterDataPipe and ToBytesIterDataPipe to the datapipes."


Summary: Add two new DataPipes, one for getting web file urls to yield streams and one for getting streams to yield bytes.

Test Plan:Add test_web_iterable_datapipe in test/test_datapipes.py. The test initiates a local http server for serving test files. test is ok.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "[Bootcamp][pytorch]Add WebIterDataPipe and ToBytesIterDataPipe to the datapipes."


Summary: Add two new DataPipes, one for getting web file urls to yield streams and one for getting streams to yield bytes.

Test Plan:Add test_web_iterable_datapipe in test/test_datapipes.py. The test initiates a local http server for serving test files. test is ok.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "[Bootcamp][pytorch]Add WebIterDataPipe and ToBytesIterDataPipe to the datapipes."


Summary: Add two new DataPipes, one for getting web file urls to yield streams and one for getting streams to yield bytes.

Test Plan:Add test_web_iterable_datapipe in test/test_datapipes.py. The test initiates a local http server for serving test files. test is ok.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jg2912 commented:
@jg2912 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29051186).

ejguan commented on test/test_datapipe.py:
```diff
@@ -77,10 +140,14 @@ def setUp(self):
         self.temp_sub_dir = ret[1][0]
         self.temp_sub_files = ret[1][1:]
 
+        set_up_local_server()
+        create_files_to_web_temp()
```
Can you create a separate TestCase class, or just move these methods into your test methods? IIRC, every single test method will call setup and tearDown, which should not be expected and too redundant in this case.

VitalyFedyunin replied:
Separate TestCase class with proper seUp and tearDown is the best here

jg2912 replied:
thank you for all comments! will change as suggested

jg2912 replied:
fyi it is now moved to a new separate testcase TestIterableDataPipeHttp.

ejguan commented on torch/utils/data/datapipes/iter/readlinesfromfile.py:
```diff
@@ -0,0 +1,18 @@
+        filepath : file path
+    """
+
+    def __init__(self, filepath):
+        self.filepath = filepath
```
Can we take a prior datapipe like LoadFilesFromDisk or ListDirFiles as input rather than one file path?

jg2912 replied:
fyi yes, it now takes an iterdatapipe<string>

ejguan commented on torch/utils/data/datapipes/iter/tobytes.py:
```diff
@@ -0,0 +1,21 @@
+from torch.utils.data import IterDataPipe
+
+
+class ToBytesIterDataPipe(IterDataPipe[Tuple[str, bytes]]):
```
Why we need this DataPipe to read bytes with a maximum limit?

ejguan replied:
IMO, it makes more sense to have a Datapipe to read specific bytes in each iteration. Then close the stream when deplete it.

jg2912 replied:
fyi, agree -- I change the parameter to be 'chunk' which is the size limit (def=None) on each read iteration. Per Vitaly, no close stream, rely on ref counter here.

ejguan commented on torch/utils/data/datapipes/iter/web.py:
```diff
@@ -0,0 +1,34 @@
+                r = urllib.urlopen(furl, timeout=self.timeout)
+
+                yield(furl, r)
+            except HTTPError as e:
+                raise Exception("Could not get the file.\
+                                [HTTP Error] {code}: {reason}."
+                                .format(code=e.code, reason=e.reason))
+            except URLError as e:
+                raise Exception("Could not get the file at {url}.\
+                                 [URL Error] {reason}."
+                                .format(reason=e.reason, url=furl))
```
These two can be combined as:
```py
except (HTTPError, URLError) as e:
    ...
```

jg2912 replied:
I am thinking of keeping them separated because they stand for different issues. Additionally HttpError has errorcode while URLError has not.

ejguan commented on torch/utils/data/datapipes/iter/web.py:
```diff
@@ -0,0 +1,34 @@
+from torch.utils.data import IterDataPipe
+
+
+class WebIterDataPipe(IterDataPipe[Tuple[str, IOBase]]):
```
Can we change name to HTTPReader? But, relying on @VitalyFedyunin about the naming of DataPipe.

VitalyFedyunin replied:
Agree about naming

jg2912 replied:
fyi, cool, name changed

ejguan commented on torch/utils/data/datapipes/iter/web.py:
```diff
@@ -0,0 +1,34 @@
+        timeout : timeout for web request
+    """
+
+    def __init__(self, source_datapipe, timeout=30):
```
Can `timeout` be `None` as default to mimic the default behavior of `urllib`?

jg2912 replied:
fyi, yes agree. It is changed to None.

ejguan commented on test/test_datapipe.py:
```diff
@@ -35,6 +41,63 @@
+        time.sleep(3)
+        print("Server started ... at {host}:{port}".format(host=HOST, port=PORT))
+    except OSError as e:
+        if e.errno == errno.EADDRINUSE:
+            # localhost:port already in use, no need to initialize again
+            print("There is already an active server serving at {host}:{port}".format(host=HOST, port=PORT))
```
Since this method will be invoked for every single method in the TestCase, you will encounter a mount of times of this Error. So, you can use `setUpClass` and `tearDownClass` to run one time per class. And I still think it's better to put them into a separate TestCase, since the server doesn't need to be created if you don't run this test method.
And, please call join for the server thread in the `tearDownClass`.

jg2912 replied:
you are right. that setup is called each time a unittest is run. 
fyi it is now moved to a new separate testcase TestIterableDataPipeHttp.

ejguan reviewed (changes requested):
Thanks for adding these DataPipe! Few comments below.

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -35,6 +41,63 @@
 
 T_co = TypeVar('T_co', covariant=True)
 
+PORT = 8000
```
nit: Generate randomly or use tool to detect free one.

jg2912 replied:
fyi, now it uses port=0 or socketserver.TCPServer(("", 0), Handler) so it automatically picks an unused port.

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -35,6 +41,63 @@
+
+        # Wait a bit for the server to come up
+        time.sleep(3)
+        print("Server started ... at {host}:{port}".format(host=HOST, port=PORT))
```
Please drop all prints

jg2912 replied:
yes, done

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -35,6 +41,63 @@
+PORT = 8000
+HOST = 'localhost'
+WEB_TEST_FILE_SIZE = 1024
+WEB_FILE_TEMP_DIR = tempfile.TemporaryDirectory(dir=os.getcwd())
```
use 
```with tempfile.TemporaryDirectory() as tmpdirname``` pattern, it will guarantee files cleanup

jg2912 replied:
fyi, yes, it now follows this tempfile usage pattern in the new test case TestIterableDataPipeHttp

VitalyFedyunin commented on torch/utils/data/datapipes/iter/tobytes.py:
```diff
@@ -0,0 +1,21 @@
+    def __iter__(self):
+        for (furl, stream) in self.source_datapipe:
+            d = stream.read(self.max_limit)
+            stream.close()
```
Please do not close stream, leave it to python's ref counters.

jg2912 replied:
got it. stream.close() is removed, relying totally on python's ref counters.
I test locally TestIterableDataPipeHttp.test_stress_http_reader_iterable_datapipes with 16M files and no issue is found. Given the default Linux IO handles max is 2M, if there is any unreleased OS handles, it would have raised the issue in the test.

VitalyFedyunin reviewed (changes requested):
Overall looks good, but requires some changes. Please also test locally on extremely large files and on large amount of files.

Commit:
Update on "[Bootcamp][pytorch]Add WebIterDataPipe and ToBytesIterDataPipe to the datapipes."


Summary: Add two new DataPipes, one for getting web file urls to yield streams and one for getting streams to yield bytes.

Test Plan:Add test_web_iterable_datapipe in test/test_datapipes.py. The test initiates a local http server for serving test files. Test below locally ok. 
1. create and load 100k localhost file urls (each of size 1KB) 
2. create and load 10k localhost file urls (each of size 1MB) 
in the unit test, for sake of testing time, put local file urls at a small number.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29051186](https://our.internmc.facebook.com/intern/diff/D29051186)

[ghstack-poisoned]

Commit:
Update on "[Bootcamp][pytorch]Add WebIterDataPipe and ToBytesIterDataPipe to the datapipes."


Summary: Add three new DataPipes, 
readlinesfromfile for getting file names to yield text lines
httpreader for getting web file urls to yield streams 
tobytes for getting streams to yield bytes.

Test Plan: new test case TestIterableDataPipeHttp in test/test_datapipes.py. The test initiates a local http server for serving test files. Local tests are ok. 
1. Create 16M files and load from the local server. Each of size 10 Bytes. Purpose is to see if any unreleased OS handles from data pipes. Notes, it takes about 5 hours to run at this configuration.
2. Create a 64GB file and load from the local server. Read is chunked at 8MB. Purpose is to see if it could load a large web file using the data pipes. 
In case of users inadvertently running into a long test case, configurations of both are put to smaller numbers.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29051186](https://our.internmc.facebook.com/intern/diff/D29051186)

[ghstack-poisoned]

Commit:
Update on "[Bootcamp][pytorch]Add WebIterDataPipe and ToBytesIterDataPipe to the datapipes."


Summary: Add three new DataPipes, 
readlinesfromfile for getting file names to yield text lines
httpreader for getting web file urls to yield streams 
tobytes for getting streams to yield bytes.

Test Plan: new test case TestIterableDataPipeHttp in test/test_datapipes.py. The test initiates a local http server for serving test files. Local tests are ok. 
1. Create 16M files and load from the local server. Each of size 10 Bytes. Purpose is to see if any unreleased OS handles from data pipes. Notes, it takes about 5 hours to run at this configuration.
2. Create a 64GB file and load from the local server. Read is chunked at 8MB. Purpose is to see if it could load a large web file using the data pipes. 
In case of users inadvertently running into a long test case, configurations of both are put to smaller numbers.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29051186](https://our.internmc.facebook.com/intern/diff/D29051186)

[ghstack-poisoned]

Commit:
Update on "[Bootcamp][pytorch]Add WebIterDataPipe and ToBytesIterDataPipe to the datapipes."


Summary: Add three new DataPipes, 
readlinesfromfile for getting file names to yield text lines
httpreader for getting web file urls to yield streams 
tobytes for getting streams to yield bytes.

Test Plan: new test case TestIterableDataPipeHttp in test/test_datapipes.py. The test initiates a local http server for serving test files. Local tests are ok. 
1. Create 16M files and load from the local server. Each of size 10 Bytes. Purpose is to see if any unreleased OS handles from data pipes. Notes, it takes about 5 hours to run at this configuration.
2. Create a 64GB file and load from the local server. Read is chunked at 8MB. Purpose is to see if it could load a large web file using the data pipes. 
In case of users inadvertently running into a long test case, configurations of both are put to smaller numbers.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29051186](https://our.internmc.facebook.com/intern/diff/D29051186)

[ghstack-poisoned]

Commit:
Update on "[Bootcamp][pytorch]Add WebIterDataPipe and ToBytesIterDataPipe to the datapipes."


Summary: Add three new DataPipes, 
readlinesfromfile for getting file names to yield text lines
httpreader for getting web file urls to yield streams 
tobytes for getting streams to yield bytes.

Test Plan: new test case TestIterableDataPipeHttp in test/test_datapipes.py. The test initiates a local http server for serving test files. Local tests are ok. 
1. Create 16M files and load from the local server. Each of size 10 Bytes. Purpose is to see if any unreleased OS handles from data pipes. Notes, it takes about 5 hours to run at this configuration.
2. Create a 64GB file and load from the local server. Read is chunked at 8MB. Purpose is to see if it could load a large web file using the data pipes. 
In case of users inadvertently running into a long test case, configurations of both are put to smaller numbers.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29051186](https://our.internmc.facebook.com/intern/diff/D29051186)

[ghstack-poisoned]

Commit:
Update on "[Bootcamp][pytorch]Add WebIterDataPipe and ToBytesIterDataPipe to the datapipes."


Summary: Add three new DataPipes, 
readlinesfromfile for getting file names to yield text lines
httpreader for getting web file urls to yield streams 
tobytes for getting streams to yield bytes.

Test Plan: new test case TestIterableDataPipeHttp in test/test_datapipes.py. The test initiates a local http server for serving test files. Local tests are ok. 
1. Create 16M files and load from the local server. Each of size 10 Bytes. Purpose is to see if any unreleased OS handles from data pipes. Notes, it takes about 5 hours to run at this configuration.
2. Create a 64GB file and load from the local server. Read is chunked at 8MB. Purpose is to see if it could load a large web file using the data pipes. 
In case of users inadvertently running into a long test case, configurations of both are put to smaller numbers.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29051186](https://our.internmc.facebook.com/intern/diff/D29051186)

[ghstack-poisoned]

Commit:
Update on "[Bootcamp][pytorch]Add WebIterDataPipe and ToBytesIterDataPipe to the datapipes."


Summary: Add three new DataPipes, 
readlinesfromfile for getting file names to yield text lines
httpreader for getting web file urls to yield streams 
tobytes for getting streams to yield bytes.

Test Plan: new test case TestIterableDataPipeHttp in test/test_datapipes.py. The test initiates a local http server for serving test files. Local tests are ok. 
1. Create 16M files and load from the local server. Each of size 10 Bytes. Purpose is to see if any unreleased OS handles from data pipes. Notes, it takes about 5 hours to run at this configuration.
2. Create a 64GB file and load from the local server. Read is chunked at 8MB. Purpose is to see if it could load a large web file using the data pipes. 
In case of users inadvertently running into a long test case, configurations of both are put to smaller numbers.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29051186](https://our.internmc.facebook.com/intern/diff/D29051186)

[ghstack-poisoned]

jg2912 commented:
@jg2912 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29051186).

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -248,6 +255,124 @@ def test_groupbykey_iterable_datapipe(self):
+                else:
+                    self.assertEqual(len(data), test_file_size)
+
+    def test_stress_http_reader_iterable_datapipes(self):
```
Please disable this test, as we can't afford to run it in CI env.

jg2912 replied:
agree, they could be slow if config to a larger number.  will disable both for now.

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -248,6 +255,124 @@ def test_groupbykey_iterable_datapipe(self):
+        test_file_count = 1024
+        self.http_test_base(test_file_size, test_file_count)
+
+    def test_large_files_http_reader_iterable_datapipes(self):
```
Please also disable this test (leave code for future)

VitalyFedyunin reviewed (changes requested):
Everything good, but please disable slow tests

Commit:
Update on "[Bootcamp][pytorch]Add WebIterDataPipe and ToBytesIterDataPipe to the datapipes."


Summary: Add three new DataPipes, 
readlinesfromfile for getting file names to yield text lines
httpreader for getting web file urls to yield streams 
tobytes for getting streams to yield bytes.

Test Plan: new test case TestIterableDataPipeHttp in test/test_datapipes.py. The test initiates a local http server for serving test files. Local tests are ok. 
1. Create 16M files and load from the local server. Each of size 10 Bytes. Purpose is to see if any unreleased OS handles from data pipes. Notes, it takes about 5 hours to run at this configuration.
2. Create a 64GB file and load from the local server. Read is chunked at 8MB. Purpose is to see if it could load a large web file using the data pipes. 
In case of users inadvertently running into a long test case, configurations of both are put to smaller numbers.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29051186](https://our.internmc.facebook.com/intern/diff/D29051186)

[ghstack-poisoned]

jg2912 commented:
@jg2912 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29051186).

ejguan commented on test/test_datapipe.py:
```diff
@@ -248,6 +256,128 @@ def test_groupbykey_iterable_datapipe(self):
+
+            datapipe_dirf = dp.iter.ListDirFiles(tmpdir, '*_list')
+            datapipe_rf = dp.iter.ReadLinesFromFile(datapipe_dirf)
+            datapipe_tup = TupleToElementIDP(datapipe_rf, 1)
```
If it's required, can you use map datapipe?

jg2912 replied:
sure -- let me quickly change it

jg2912 replied:
I replace TupleToElementIDP with MapIterDataPipe while keeping the explicit pipe stack of ListDirFiles and ReadLinesFromFile. My thinking is TupleToElementIDP is more temporary, used specifically for unittest. The other two could follow the pattern of TestIterableDataPipeBasic.

ejguan replied:
It can be simplified as:
```py
def _get_from_tuple(data):
    return data[1]
datapipe_tup = dp.iter.Map(datapipe_f_lines, _get_from_tuple)
```

jg2912 replied:
cool. then my proposed change is quite similar with yours

def _get_data_from_tuple_fn(data, *args, **kwargs):
            return data[args[0]]
...
datapipe_line_url = dp.iter.Map(datapipe_f_lines,
                                _get_data_from_tuple_fn, [1])

Commit:
Update on "[Bootcamp][pytorch]Add HTTPReaderIterDataPipe and ToBytesIterDataPipe"


Summary: Add three new DataPipes, 
readlinesfromfile for getting file names to yield text lines
httpreader for getting web file urls to yield streams 
tobytes for getting streams to yield bytes.

Test Plan: new test case TestIterableDataPipeHttp in test/test_datapipes.py. The test initiates a local http server for serving test files. Local tests are ok. 
1. Create 16M files and load from the local server. Each of size 10 Bytes. Purpose is to see if any unreleased OS handles from data pipes. Notes, it takes about 5 hours to run at this configuration.
2. Create a 64GB file and load from the local server. Read is chunked at 8MB. Purpose is to see if it could load a large web file using the data pipes. 
In case of users inadvertently running into a long test case, configurations of both are put to smaller numbers.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29051186](https://our.internmc.facebook.com/intern/diff/D29051186)

[ghstack-poisoned]

Commit:
Update on "[Bootcamp][pytorch]Add HTTPReaderIterDataPipe and ToBytesIterDataPipe"


Summary: Add three new DataPipes, 
readlinesfromfile for getting file names to yield text lines
httpreader for getting web file urls to yield streams 
tobytes for getting streams to yield bytes.

Test Plan: new test case TestIterableDataPipeHttp in test/test_datapipes.py. The test initiates a local http server for serving test files. Local tests are ok. 
1. Create 16M files and load from the local server. Each of size 10 Bytes. Purpose is to see if any unreleased OS handles from data pipes. Notes, it takes about 5 hours to run at this configuration.
2. Create a 64GB file and load from the local server. Read is chunked at 8MB. Purpose is to see if it could load a large web file using the data pipes. 
In case of users inadvertently running into a long test case, configurations of both are put to smaller numbers.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29051186](https://our.internmc.facebook.com/intern/diff/D29051186)

[ghstack-poisoned]

ejguan commented on test/test_datapipe.py:
```diff
@@ -248,6 +256,134 @@ def test_groupbykey_iterable_datapipe(self):
+                           not tear down (clean up temp directory or terminate\
+                           local server) due to {0}".format(str(e)))
+
+    def http_test_base(self, test_file_size, test_file_count, timeout=None,
```
BTW, if it's a helper function, please add an underscore.

jg2912 replied:
agree, fixed

Commit:
Update on "[Bootcamp][pytorch]Add HTTPReaderIterDataPipe and ToBytesIterDataPipe"


Summary: Add three new DataPipes, 
readlinesfromfile for getting file names to yield text lines
httpreader for getting web file urls to yield streams 
tobytes for getting streams to yield bytes.

Test Plan: new test case TestIterableDataPipeHttp in test/test_datapipes.py. The test initiates a local http server for serving test files. Local tests are ok. 
1. Create 16M files and load from the local server. Each of size 10 Bytes. Purpose is to see if any unreleased OS handles from data pipes. Notes, it takes about 5 hours to run at this configuration.
2. Create a 64GB file and load from the local server. Read is chunked at 8MB. Purpose is to see if it could load a large web file using the data pipes. 
In case of users inadvertently running into a long test case, configurations of both are put to smaller numbers.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29051186](https://our.internmc.facebook.com/intern/diff/D29051186)

[ghstack-poisoned]

jg2912 commented:
@jg2912 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29051186).

Commit:
Update on "[Bootcamp][pytorch]Add HTTPReaderIterDataPipe and ToBytesIterDataPipe"


Summary: Add three new DataPipes, 
readlinesfromfile for getting file names to yield text lines
httpreader for getting web file urls to yield streams 
tobytes for getting streams to yield bytes.

Test Plan: new test case TestIterableDataPipeHttp in test/test_datapipes.py. The test initiates a local http server for serving test files. Local tests are ok. 
1. Create 16M files and load from the local server. Each of size 10 Bytes. Purpose is to see if any unreleased OS handles from data pipes. Notes, it takes about 5 hours to run at this configuration.
2. Create a 64GB file and load from the local server. Read is chunked at 8MB. Purpose is to see if it could load a large web file using the data pipes. 
In case of users inadvertently running into a long test case, configurations of both are put to smaller numbers.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29051186](https://our.internmc.facebook.com/intern/diff/D29051186)

[ghstack-poisoned]

jg2912 commented:
@jg2912 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29051186).

Commit:
Update on "[Bootcamp][pytorch]Add HTTPReaderIterDataPipe and ToBytesIterDataPipe"


Summary: Add three new DataPipes, 
readlinesfromfile for getting file names to yield text lines
httpreader for getting web file urls to yield streams 
tobytes for getting streams to yield bytes.

Test Plan: new test case TestIterableDataPipeHttp in test/test_datapipes.py. The test initiates a local http server for serving test files. Local tests are ok. 
1. Create 16M files and load from the local server. Each of size 10 Bytes. Purpose is to see if any unreleased OS handles from data pipes. Notes, it takes about 5 hours to run at this configuration.
2. Create a 64GB file and load from the local server. Read is chunked at 8MB. Purpose is to see if it could load a large web file using the data pipes. 
In case of users inadvertently running into a long test case, configurations of both are put to smaller numbers.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29051186](https://our.internmc.facebook.com/intern/diff/D29051186)

[ghstack-poisoned]

jg2912 commented:
@jg2912 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29051186).
